src/client/async_forward_backward.py

```diff
@@ -0,0 +1,215 @@
+import logging
+from typing import Optional, Union, List, Sequence, Tuple, Dict
+
+import torch
+from hivemind import DHT, P2P, get_logger, use_hivemind_log_handler
+from hivemind.moe.client.remote_expert_worker import RemoteExpertWorker
+from torch import nn
+
+import asyncio
+from src.client.sequence_manager import RemoteSequenceManager
+
+from hivemind import (
+    P2P,
+    get_logger,
+    nested_flatten,
+    serialize_torch_tensor,
+    use_hivemind_log_handler,
+)
+
+from hivemind.utils.nested import nested_compare, nested_flatten, nested_pack
+from hivemind.moe.client.expert import expert_forward, expert_backward
+from hivemind.moe.client.remote_expert_worker import RemoteExpertWorker
+from hivemind.p2p import StubBase
+
+from src.client.sequence_manager import RemoteSequenceManager
+from src.server.handler import TransformerConnectionHandler
+from src.data_structures import CHAIN_DELIMITER, RemoteSpanInfo, ModuleUID, RemoteSpanInfo, RPCInfo
+
+
+MAX_TOKENS_IN_BATCH=1024
+
+
+async def run_forward(
+    uid: ModuleUID, 
+    stub: StubBase,
+    rpc_info: RPCInfo,
+    *inputs: torch.Tensor,
+    **kwargs
+) -> Tuple[torch.Tensor, ...]:
+    # Note: *inputs are flattened input tensors that follow the expert's info['input_schema']
+    # detach to avoid pickling the computation graph
+    assert len(kwargs) == len(rpc_info["keyword_names"]), f"Keyword args should be {rpc_info['keyword_names']}"
+    kwargs = {key: kwargs[key] for key in rpc_info["keyword_names"]}
+    
+    # Note: we put keyword arguments in the same order as on a server to prevent f(a=1, b=2) != f(b=2, a=1) errors
+    forward_inputs = (inputs, kwargs)
+
+    if not nested_compare(forward_inputs, rpc_info["forward_schema"]):
+        raise TypeError(f"Inputs do not match expert input schema. Did you pass the right number of parameters?")
+ 
+    forward_inputs = nested_flatten(forward_inputs)
+    inputs = tuple(tensor.cpu().detach() for tensor in forward_inputs)
+
+    serialized_tensors = (
+        serialize_torch_tensor(tensor, proto.compression)
```

Comment:
[note to self: create a todo to figure out whether we should use run_in_executor]